### PR TITLE
URIParser and removed `transformer.transform` from trainer (as well as refactors)

### PR DESCRIPTION
**Pull request by richardwu**
Wanted to first refactor out this https://github.com/richardwu/news_scraper/compare/uri_parser_and_trainer_refactor_1?expand=1#diff-1ee27f6b90d971e3cdcf966d679cfb11L19 since we don't really need it once we discover that a domain is not in the config file.

Ended up having to refactor and fix a lot of tests because of how much we depended on our private interface.

I think a good next step is to move some of the helper methods (e.g. `no_scrape_defined`, `save_selected_presets`) into private methods and do a full integration test using the public interface(i.e. `UriTrainer#train` and `PresetsSelector.select`) instead. 

cc: @jules2689

### CODE REVIEW

File: lib/news_scraper/trainer/uri_trainer.rb [hunk 1]
```diff
@@ -3,60 +3,53 @@ module Trainer
...
+          no_scrape_defined
         end
       end
 
-      def no_scrape_defined(root_domain)
-        if CLI.confirm("Step through presets for #{root_domain}?")
+      def no_scrape_defined
+        if CLI.confirm("Step through presets for #{@root_domain}?")
           CLI.put_footer
-          selected_presets = @data_type.train(@raw_html)
```
Comment (richardwu) [inline lib/news_scraper/trainer/uri_trainer.rb:31]: This line actually caused a lot of trouble because our tests stubbed out `PresetsSelector#train` e.g. https://github.com/richardwu/news_scraper/pull/3/files#diff-1f9d04374200e0405dea75c16e01751fL47.

This goes to show that testing these individual methods doesn't necessarily improve our test coverage.

In fact, the stubbing caused us to miss this mistake.

I think we need to refactor our tests and public/private interface so that we get true test coverage for our public interface.

File: lib/news_scraper/cli.rb [hunk 2]
```diff
@@ -83,7 +83,7 @@ def put_edge(color, prefix, text)
       ptext = "#{color}#{prefix}#{text}"
       textwidth = printing_width(ptext)
 
-      termwidth = IO.console ? IO.console.winsize[1] : 80
+      termwidth = IO.respond_to?(:console) && IO.console ? IO.console.winsize[1] : 80
```
Comment (jules2689) [inline lib/news_scraper/cli.rb:86]: Why is this needed?

File: lib/news_scraper/uri_parser.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+module NewsScraper
+  class URIParser
```
Comment (jules2689) [inline lib/news_scraper/uri_parser.rb:2]: Why not just use Ruby's `URI`?
`host` => `URI.host` for example, also why do we remove the scheme for the HTTP request? Half the servers will just redirect

**Comment by jules2689** [comment]
Changes make sense, `URIParser` should just be ruby's `URI` though. Have the question about the schemeless URI stuff.

I'll take a stab at really refactoring this after we merge this through. I hadn't taken a deep look at what we had in these files past splitting up functionality into multiple files - it's otherwise original code. We're at the point where we can really look at what it does and change it up to match the intended use case much better.

File: lib/news_scraper/cli.rb (hunk 2 above)
Comment (richardwu) [reply lib/news_scraper/cli.rb:86]: `pry` reassigns `IO` as their own `IO` object which doesn't have `.console`

File: lib/news_scraper/uri_parser.rb (hunk 3 above)
Comment (richardwu) [reply lib/news_scraper/uri_parser.rb:2]: We'll need to do some funky check to see if the scheme already exists if we want to use `URI.host` http://stackoverflow.com/a/6674363/4463793.

I also wanted to write my own regex and this is much more succinct in my opinion.

We need to filter the scheme (and then add `http://` https://github.com/richardwu/news_scraper/blob/master/lib/news_scraper/extractors/article.rb#L15) before sticking it into `HTTParty` because it'll try to connect via pure TCP without HTTP, and some urls may not come prepended with `http://`.

**Comment by richardwu** [comment]
Ruby's URI is pretty garbage. There's a lot of specific stuff you have to
handle (not to mention their "host" method doesn't remove the TLD) it's
worth having a wrapper around it for our specific use cases.
